File: leetcode/26-remove-duplicates-from-sorted-array/26-remove-duplicates-from-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
  let count = 0;
  let current = null;

  for (let i = 0; i < nums.length; i++) {
    if (current !== nums[i]) {
      nums[count] = nums[i];
      count++;
      current = nums[i];
    }
  }
  return count;
};


// test function
const test = function(q, a) {
  const sol = removeDuplicates(q);
  console.assert(sol === a.length, `Fail: sol ${sol}, length ${a.length}`);
  for (let i = 0; i < sol; i++) {
    console.assert(a[i] === q[i], `Fail: ${a[i]} !== ${q[i]}`);
  }
};


const q1 = [1, 1, 2];
const a1 = [1, 2];
test(q1, a1);

const q2 = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4];
const a2 = [0, 1, 2, 3, 4];
test(q2, a2);

File: LeetCode/26-remove-duplicates-from-sorted-array/26-remove-duplicates-from-sorted-array.py
from dataclasses import dataclass


# class Solution:
#     def removeDuplicates(self, nums: 'list[int]') -> 'int':
#         count = 0
#         current = None
#         for num in nums:
#             if current != num:
#                 nums[count] = num
#                 count += 1
#                 current = num
#
#         return count


# class Solution:
#     def removeDuplicates(self, nums: 'list[int]') -> 'int':
#         count = 0
#         current = None
#         for _, num in enumerate(nums):
#             if current != num:
#                 nums[count] = num
#                 count += 1
#                 current = num
#         return count


# class Solution:
#     def removeDuplicates(self, nums: 'list[int]') -> 'int':
#         count = 0
#         current = None
#         i = 0
#         while i < len(nums):
#             if current != nums[i]:
#                 nums[count] = nums[i]
#                 count += 1
#                 current = nums[i]
#             i += 1
#         return count

class Solution:
    def removeDuplicates(self, nums: 'list[int]') -> 'int':
        current = nums[0]
        i = 1
        j = 1
        while i < len(nums):
            if current != nums[i]:
                nums[j] = nums[i]
                j += 1
                current = nums[i]
            i += 1
        return j


@dataclass
class Quest():
    inp: list
    sol: list


def checkAns(q: Quest) -> bool:
    sol = Solution()
    out = sol.removeDuplicates(q.inp)
    res = (q.sol == q.inp[:out])
    if not res:
        print(f'{q.sol} != {q.inp}')
    return res


if __name__ == '__main__':
    q = Quest(inp=[1, 1, 2], sol=[1, 2])
    checkAns(q)

    q = Quest(inp=[0, 0, 1, 1, 1, 2, 2, 3, 3, 4], sol=[0, 1, 2, 3, 4])
    checkAns(q)

    q = Quest(inp=[0], sol=[0])
    checkAns(q)
